doc/design/cpp_data_feeding.md

```diff
@@ -1,53 +1,97 @@
 # C++ Data Feeding
 
-While using Paddle V2 API for Training, data feeding completely depends on the Python code. To get rid of the Python environment and achieve the goal of "wrapping the whole training by a while loop op" in Paddle Fluid, a C++ data feeding mechanism is required. 
+While using Paddle V2 API for training, data feeding completely depends on the Python code. To get rid of the Python environment and achieve the goal of "wrapping the whole training by a while loop op" in Paddle Fluid, a C++ data feeding mechanism is required.
 
-In this document we show the fundamental design of a C++ data feeding process, which includes data reading, shuffling and batching.
+In this document, we show the fundamental design of a C++ data feeding process, which includes data reading, shuffling and batching.
+
+## Overview
+
+![](images/readers.png)
 
 ## Reader
 
-In order to handle the above mentioned problem, a new concept called 'Reader' is introduced. `Reader` is a series of inherited classes which can be held by our `Variable` and they are used to read or process file data.
+In order to handle the above-mentioned problem, a new concept called 'Reader' is introduced. `Reader` is a series of inherited classes which can be held by our `Variable` and they are used to read or process file data.
 
 
-### `ReaderBase`
+### ReaderBase
 
 `ReaderBase` is the abstract base class for all readers. It defines the interface for all readers.
 
 ```cpp
 class ReaderBase {
  public:
-  explicit ReaderBase(const std::vector<DDim>& shapes) : shapes_(shapes) {
-    PADDLE_ENFORCE(!shapes_.empty());
-  }
-  // Read the next batch of data. (A 'batch' can be only one instance)
-  // If the next batch doesn't exist, '*out' will be an empty std::vector.
+  // Reads the next batch of data. (A 'batch' can be only one instance)
+  // If the next batch doesn't exist, it throws an exception
   virtual void ReadNext(std::vector<LoDTensor>* out) = 0;
   
-  // Reinitialize the reader and read the file from the beginning.
-  virtual void ReInit() = 0;
+  // Checks whether the next instance exists.
+  virtual bool HasNext() = 0;
```

Comment:
Excellent!

Reply:
Of course we can return `EOF`. We do need to do something when users try to invoke `ReadNext()` while there is no next data. It can be throwing an exception or returning an `EOF`. Both one is OK.

But I'm afraid the `HasNext()` is still needed. Sometimes users would like to do something(like saving model) after each pass training. It requires an interface to check whether a pass is completed.

Maybe we can also implement the checking by trying invoking `ReadNext()` and check the return value. But it seems a little hard.